### PR TITLE
Added some basic instrumentation for the publisher/outputs

### PR DESCRIPTION
**Pull request by tsg**
The following metrics are exported via expvars (i.e. -httpprof interface):

* libbeatPublishedEvents - For how many events PublishEvent or PublishEvents
  are called. I.e. how many events the Beat is creating.
* libbeatMessagesInWorkerQueues - How many messages (note that a message can
  have lots of events inside - default 10k) are sitting in the publisher worker
  queues.
* libbeatMessagesDropped - How many messages are dropped by the outputs using
  one of the load balancing, single or failover modes
* libbeatEsPublishedAndAckedEvents - How many events were ack-ed by Elasticsearch
* libbeatLogstashPublishedAndAckedEvents - How many events were ack-ed by
  Logstash

### CODE REVIEW

File: libbeat/outputs/elasticsearch/client.go [hunk 1]
```diff
@@ -104,9 +110,10 @@ func (client *Client) PublishEvents(
 	}
 
 	// check response for transient errors
-	events = bulkCollectPublishFails(res, events)
-	if len(events) > 0 {
-		return events, mode.ErrTempBulkFailure
+	failed_events := bulkCollectPublishFails(res, events)
+	ackedEvents.Add(int64(len(events) - len(failed_events)))
+	if len(failed_events) > 0 {
+		return failed_events, mode.ErrTempBulkFailure
```
Comment (urso) [inline libbeat/outputs/elasticsearch/client.go]: some additional metrics might be of interest:
- number of PublishEvents calls
- number of nonACKed events

File: libbeat/outputs/logstash/client.go [hunk 2]
```diff
@@ -133,6 +139,8 @@ func (l *lumberjackClient) publishWindowed(events []common.MapStr) (int, error)
 		}
 	}
 
+	ackedEvents.Add(int64(len(events)))
```
Comment (urso) [inline libbeat/outputs/logstash/client.go]: better do counting in PublishEvents. this function returns number of events being send and ACKed

File: libbeat/outputs/mode/balance.go [hunk 3]
```diff
@@ -231,6 +233,7 @@ func (m *LoadBalancerMode) onFail(msg eventsMessage, err error) {
 	logp.Info("Error publishing events (retrying): %s", err)
 
 	if ok := m.forwardEvent(m.retries, msg); !ok {
+		messagesDropped.Add(1)
```
Comment (urso) [inline libbeat/outputs/mode/balance.go]: maybe add a method 'dropping(msg eventsMessage)' to LoadBalancerMode, so we have less repetitive code and add more metrics at this place later if required.

File: libbeat/outputs/mode/failover.go [hunk 4]
```diff
@@ -183,6 +183,7 @@ func (f *FailOverConnectionMode) publish(
 		}
 		if f.maxAttempts > 0 && fails == f.maxAttempts {
 			// max number of attempts reached
+			messagesDropped.Add(1)
```
Comment (urso) [inline libbeat/outputs/mode/failover.go]: can be moved to line 193. Having this close to SignalFailed (plus put in another method) better clarifies these to belonging together.

File: libbeat/outputs/mode/single.go [hunk 5]
```diff
@@ -137,6 +137,7 @@ func (s *SingleConnectionMode) publish(
 		}
 		if s.maxAttempts > 0 && fails == s.maxAttempts {
 			// max number of attempts reached
+			messagesDropped.Add(1)
```
Comment (urso) [inline libbeat/outputs/mode/single.go]: same here. move closer to SignalFailed

**Comment by tsg** [comment]
@urso I think I address all your comments, can you have a look again, please?

**Comment by dedemorton** [comment]
@tsg Wondering if we want to say anything  about these metrics in the topic(s) where we describe the command line option: https://www.elastic.co/guide/en/beats/libbeat/master/command-line-options.html. If you think that what we have in the doc is good enough, then please disregard my comment.

**Comment by tsg** [comment]
@dedemorton it's good enough already, we don't want to document this stuff, it's just for our own troubleshooting.

**Pull request merged**